### PR TITLE
[BB-3375] Add boto3 as individual dependency

### PR DESCRIPTION
**Pull request by farhaanbukhsh**
For  now boto3 comes as a dependency of django-ses, but it is as well require to make use of S3 storage medium class.
Especially if you want to store media in S3 bucket. Something like:

```

DISCOVERY_MEDIA_STORAGE_BACKEND:
  DEFAULT_FILE_STORAGE: 'storages.backends.s3boto3.S3Boto3Storage'

```

**JIRA tickets**: Mention Jira ticket(s) here.

~~**Screenshots**: Always include screenshots if there is any change to the UI. Can be useful for people that are not reviewer but want to know what's going on.~~

~~**Sandbox URL**: TBD - sandbox is being provisioned (if needed).~~

**Testing instructions**:

Just dependency addition

**Author notes and concerns**:

This is just to make sure to use the latest storage with S3

**Reviewers**
- [ ] @shimulch

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @farhaanbukhsh! I've created [OSPR-5357](https://openedx.atlassian.net/browse/OSPR-5357) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

**Comment by natabene** [comment]
@farhaanbukhsh Thank you for your contribution. Can you please look into resolving the conflict?

**Comment by farhaanbukhsh** [comment]
@natabene I have resolved the conflicts

**Review by shimulch — APPROVED** [review]
LGTM :+1:

- [x] I read through the code
- [N/A] I checked for accessibility issues
- [N/A] Includes documentation
- [N/A] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.

**Comment by natabene** [comment]
@jmyatt Here is another OSPR for your squad, could please review?

File: requirements/production.txt [hunk 1]
```diff
@@ -4,135 +4,426 @@
...
-unicode-slugify==0.1.3    # via -r requirements/base.in
-unicodecsv==0.14.1        # via djangorestframework-csv
-unidecode==1.1.2          # via unicode-slugify
-uritemplate==3.0.1        # via coreapi
-urllib3==1.26.2           # via botocore, elasticsearch, requests
-vine==1.3.0               # via amqp, celery
-xss-utils==0.2.0          # via -r requirements/base.in
-zope.event==4.5.0         # via gevent
-zope.interface==5.2.0     # via gevent
+algoliasearch-django==1.7.2
```
Comment (mikix) [inline requirements/production.txt:7]: Whoa what happened with this formatting? Is this a different (newer?) version of `pip-compile`?

**Comment by mikix** [comment]
If a changed config is introducing an S3 dependency, why not just also install boto3 as part of those customizations?

Like... if someone out there wants to use Azure to store media, it wouldn't feel appropriate to add an azure dep in the main code.

File: requirements/production.txt (hunk 1 above)
Comment (farhaanbukhsh) [reply requirements/production.txt:7]: Hey @mikix  thanks for replying I think pip-compile is normally behaving like that I saw the PR https://github.com/edx/course-discovery/pull/2879/files which also have the same formatting.

I am using:

```
➜ pip-compile --version
pip-compile, version 5.5.0

```

Let me know if I need to use any other command or pass some option. :sweat:

**Comment by farhaanbukhsh** [comment]
> If a changed config is introducing an S3 dependency, why not just also install boto3 as part of those customizations?
> 
> Like... if someone out there wants to use Azure to store media, it wouldn't feel appropriate to add an azure dep in the main code.

@mikix that's a valid point but then how can one install selected dependency for course-discovery while deploying? Also boto3 anyhow is getting introduced through django-ses.

**Comment by mikix** [comment]
> how can one install selected dependency for course-discovery while deploying?

Good question. I'm not super familiar with the instructions for deploying open edx. But I assume there's some scripting one could add in there or something in `edx/configuration` you could override?

In our docs about it, we talk like running `pip` would be no problem during the deploy process:
https://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/install_xblock.html 

> Also boto3 anyhow is getting introduced through django-ses.

Well true. But that reminds me - why this change, if boto3 is coming through anyway?

**Comment by farhaanbukhsh** [comment]
> > how can one install selected dependency for course-discovery while deploying?
> 
> Good question. I'm not super familiar with the instructions for deploying open edx. But I assume there's some scripting one could add in there or something in `edx/configuration` you could override?
> 

I couldn't find the options when I looked through the playbook.

> In our docs about it, we talk like running `pip` would be no problem during the deploy process:
> https://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/install_xblock.html
> 
It's not scalable to manually run `pip` in each machine

> > Also boto3 anyhow is getting introduced through django-ses.
> 
> Well true. But that reminds me - why this change, if boto3 is coming through anyway?

So what we were trying to do was make `media asset ` to use a S3 backend and in order to do that we need to use `'storages.backends.s3boto3.S3Boto3Storage'` and for that boto3 was required. All of these were being done on Juniper and the version of django-ses that is pinned doesn't have boto3 dependency so we ended up adding boto3 in our fork to help the backend run smoothly.

We thought that people should not face this issue, right now it's okay because boto3 anyhow come as a side-effect.

So if we explicitly keep it out there it would be better.

And I find your point of adding the backend supporting dependency in a different way. But I couldn't find any other way. So either we add a way through edx/configuration or we explicitly mention and document what one needs to do to support different storage solution.

What do you feel about it?

Thanks a ton @mikix  for replying :smile:  means a lot.

**Comment by mikix** [comment]
> It's not scalable to manually run pip in each machine

Yeah not manually. I was just saying using pip as part of some deploy script. I honestly don't know enough about open edx deployment to be able to point to where you'd do this. But I know it's possible. edx.org installs non-default xblocks and stuff. So it should be a part of the sequence somewhere.

Maybe ask in the open edx discourse, to find someone more knowledgable than me, on how to install custom dependencies during deploy? That could clear our questions up.

**Comment by nedbat** [comment]
I don't understand all the factors at work here, but it sounds like we shouldn't add a dependency in case some operators want to use it with their customized settings.  If it's difficult to customize the set of extra packages installed for a particular service, we should work on improving that.  We won't be able to add all the dependencies that people might need.

**Comment by farhaanbukhsh** [comment]
@nedbat @mikix 

It totally makes sense to not add the dependency of a 3rd party storage service in the main repo. I struggled to find a way to install a 3rd party dependency (Aws s3 or Azure) for IDA's (course-discovery here) in edx-configuration repo as well. So if you folks know of a way to do it and if you can point out to me that will be great else we should file a ticket to add it I believe. I hope I am able convey my pain points. Thanks again for spending time on this and for answering my question. :) Means a lot to me :heart: 

I am closing this PR for now.

**Comment by openedx-webhooks** [comment]
@farhaanbukhsh Even though your pull request wasn’t merged, please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/course-discovery&entry.867055334=https://github.com/edx/course-discovery/pull/2882&entry.1484655318=https://github.com/farhaanbukhsh&entry.752974735=2021-01-12+18:20&entry.1917517419=2021-02-01+07:21&entry.2133058324=No) so we can improve your experience in the future.

**Comment by mikix** [comment]
@farhaanbukhsh unfortunately, I only ever really deal with open edx installations from a devstack perspective - I don't have too much context from an operational perspective. I think you should ask on the forums: https://discuss.openedx.org/

**Comment by nedbat** [comment]
@farhaanbukhsh I would ask in [Discourse](https://discuss.openedx.org) how people deal with these sorts of needs.

**Comment by farhaanbukhsh** [comment]
@nedbat @mikix https://discuss.openedx.org/t/how-to-install-private-dependencies-while-deploying-idas/4177

I have put up a question on discourse thanks for guiding me here :smile:

**Pull request closed without merging**